pkg/upgrade/deploy.go

```diff
@@ -2,44 +2,32 @@ package upgrade
 
 import (
 	"context"
-	"encoding/json"
-
-	"github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2018-05-01/resources"
-	"github.com/openshift/openshift-azure/pkg/util/azureclient"
 
 	"github.com/openshift/openshift-azure/pkg/api"
-	"github.com/openshift/openshift-azure/pkg/initialize"
-	"github.com/openshift/openshift-azure/pkg/log"
+	"github.com/openshift/openshift-azure/pkg/util/managedcluster"
 )
 
-func Deploy(ctx context.Context, cs *api.OpenShiftManagedCluster, i initialize.Initializer, azuredeploy []byte, config api.PluginConfig) error {
-	var t map[string]interface{}
-	err := json.Unmarshal(azuredeploy, &t)
+func (u *simpleUpgrader) Deploy(ctx context.Context, cs *api.OpenShiftManagedCluster, azuredeploy []byte, deployFn api.DeployFn) error {
+	err := deployFn(ctx, cs, azuredeploy)
 	if err != nil {
 		return err
 	}
 
-	clients, err := azureclient.NewAzureClients(ctx, cs, config)
+	err = u.InitializeCluster(ctx, cs)
 	if err != nil {
 		return err
 	}
-
-	log.Info("applying arm template deployment")
-	future, err := clients.Deployments.CreateOrUpdate(ctx, cs.Properties.AzProfile.ResourceGroup, "azuredeploy", resources.Deployment{
-		Properties: &resources.DeploymentProperties{
-			Template: t,
-			Mode:     resources.Incremental,
-		},
-	})
+	kc, err := managedcluster.ClientSetFromV1Config(ctx, cs.Config.AdminKubeconfig)
 	if err != nil {
 		return err
 	}
 
-	log.Info("waiting for arm template deployment to complete")
-	err = future.WaitForCompletionRef(ctx, clients.Deployments.Client)
+	// ensure that all nodes are ready
+	err = WaitForNodes(ctx, cs, kc)
 	if err != nil {
 		return err
 	}
 
-	return i.InitializeCluster(ctx, cs)
+	// Wait for infrastructure services to be healthy
+	return WaitForInfraServices(ctx, kc)
```

Comment:
you can remove this if the caller is calling it for you

Reply:
done